sycl/include/sycl/ext/oneapi/properties/property_utils.hpp

```diff
@@ -0,0 +1,221 @@
+//==-- property_utils.hpp --- SYCL extended property list common utilities -==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/detail/property_helper.hpp>
+#include <sycl/ext/oneapi/properties/property.hpp>
+
+#include <tuple>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+// Forward declaration
+template <typename PropertyT, typename T, typename... Ts> struct property_value;
+
+namespace detail {
+
+//******************************************************************************
+// Misc
+//******************************************************************************
+
+// Checks if a type is a tuple.
+template <typename T> struct IsTuple : std::false_type {};
+template <typename... Ts> struct IsTuple<std::tuple<Ts...>> : std::true_type {};
+
+// Gets the first type in a parameter pack of types.
+template <typename... Ts>
+using GetFirstType = typename std::tuple_element<0, std::tuple<Ts...>>::type;
+
+// Prepends a value to a tuple.
+template <typename T, typename Tuple> struct PrependTuple {};
+template <typename T, typename... Ts>
+struct PrependTuple<T, std::tuple<Ts...>> {
+  using type = std::tuple<T, Ts...>;
+};
+
+// Checks if a type T has a static value member variable.
+template <typename T, typename U = int> struct HasValue : std::false_type {};
+template <typename T>
+struct HasValue<T, decltype((void)T::value, 0)> : std::true_type {};
+
+// Checks that all types in a tuple have unique PropertyID.
+template <typename T> struct AllUnique {};
+template <typename... Ts>
+struct AllUnique<std::tuple<Ts...>> : std::true_type {};
+template <typename T> struct AllUnique<std::tuple<T>> : std::true_type {};
+template <typename L, typename R, typename... Rest>
+struct AllUnique<std::tuple<L, R, Rest...>>
+    : sycl::detail::conditional_t<PropertyID<L>::value != PropertyID<R>::value,
```

Comment:
It seems it checks that neighbors have unique ids rather than that all types have unique ids.

Reply:
You are right. This assumes that the properties have been sorted first, in which case neighbor-comparison is enough to ensure all IDs are unique. I have moved this to be among the other sorting-related traits and renamed it to `SortedAllUnique`.